Binary sensor map implementation

mvturnho opened the pull request:
## Description:


**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

mvturnho committed:
Merge pull request #1 from esphome/dev

update dev branch

mvturnho committed:
Merge pull request #2 from esphome/dev

Update dev branch

mvturnho committed:
add binary_sensor_map c code

mvturnho committed:
add python file

mvturnho committed:
fixed python and C++ code for new framework

mvturnho committed:
Merge pull request #3 from esphome/dev

update branch from source

mvturnho committed:
renamed add_sensor to add_channel

mvturnho committed:
travis

mvturnho commented:
@OttoWinter can we also include this component? I have implemented this in several applications so it is well tested.

OttoWinter commented on esphome/components/binary_sensor_map/sensor.py:
```diff
@@ -0,0 +1,38 @@
+DEPENDENCIES = ['binary_sensor']
+
+binary_sensor_map_ns = cg.esphome_ns.namespace('binary_sensor_map')
+BinarySensorMap = binary_sensor_map_ns.class_('BinarySensorMap', cg.Component)
```
This also inherits from `sensor.Sensor`

OttoWinter commented on esphome/components/binary_sensor_map/sensor.py:
```diff
@@ -0,0 +1,38 @@
+
+entry = {
+    cv.Required(CONF_CHANNEL): cv.use_id(binary_sensor.BinarySensor),
+    cv.Required(CONF_VALUE): cv.All(cv.positive_int, cv.Range(min=0, max=255)),
```
Why is this restricted to integers? The struct accepts a float.

OttoWinter commented on esphome/components/binary_sensor_map/binary_sensor_map.h:
```diff
@@ -0,0 +1,59 @@
+  std::vector<BinarySensorMapChannel *> sensors_{};
+  BinarySensorMapType sensor_type_{BINARY_SENSOR_MAP_TYPE_GROUP};
+  bool last_any_active_{false};
+  // this gives max 46 channels per binary_sensor_map
```
64, not 46 😉

OttoWinter committed:
Merge branch 'dev' into binary_sensor_map

OttoWinter committed:
Updates

- Use struct for channels_ array - heap allocation is not really necessary here.
- any_active can also be written as mask != 0
- Update setup priority to DATA
- Use shorter TAG (name is already long; not important)
- Quotes around name
- Add icon to sensor
- Use new cv.typed_schema
- Change CONF_CHANNEL to CONF_BINARY_SENSOR - makes it clearer that this option accepts a binary sensor (and not for example an int)
- Add test

OttoWinter reviewed (approved):
Yeah sure, this can get into 1.13

I have updated the PR with some things that I think could have been slightly improved (see commit message for more info).

In this state, I think it is done. I've also tested the changes with an ESP32 touch sensor and can confirm it works.
